### ISSUE TITLE
Gradients are zero when clipping values in function definition

### ISSUE THREAD

**Issue by adamvvu** [tensorflow/tensorflow#89110]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.18; 2.10

### Custom code

Yes

### Python version

3.10

### Current behavior?

I encountered a strange issue where gradients are zero when clipping-related TensorFlow Ops are defined in functions.

For context, I was implementing a numerically stable version of sigmoid where the inputs are clipped to `[-10, 10]`. However this bug can be reproduced with other functions. 

In the provided sample code, I define a truncated version of f(x) = x^2 where the function `truncate_domain` is used to clip the input tensors. However gradients only work as expected when the clipping occurs inside `GradientTape` and not when it's part of the function definition.

Reproduced on TensorFlow 2.10 with GPU, and latest 2.18 with CPU

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

def square1(x):
    return x**2

def truncate_domain(x, lb, ub):
    # Truncate for numerical stability
    x = tf.clip_by_value(x, lb, ub)
    #x = tf.math.maximum(x, lb)
    #x = tf.math.minimum(x, ub)
    return x

def square2(x):
    x = truncate_domain(x, -10., 10.)
    return x**2

def square3(x):
    x = truncate_domain(x, -10., 10.)
    return square1(x)

# Note: All 3 square functions are exactly mathematically equivalent when x in [-10, 10].

## Checking gradients

x = tf.constant(15., dtype=tf.float32)
with tf.GradientTape() as tape:
    tape.watch(x)
    x = truncate_domain(x, -10., 10.)
    l = square1(x)
grad1 = tape.gradient(l, x)
print(l)
print(grad1)
# >> tf.Tensor(100.0, shape=(), dtype=float32)
# >> tf.Tensor(20.0, shape=(), dtype=float32)

x = tf.constant(15., dtype=tf.float32)
with tf.GradientTape() as tape:
    tape.watch(x)
    l = square2(x)
grad2 = tape.gradient(l, x)
print(l)
print(grad2)
# >> tf.Tensor(100.0, shape=(), dtype=float32)
# >> tf.Tensor(0.0, shape=(), dtype=float32)

x = tf.constant(15., dtype=tf.float32)
with tf.GradientTape() as tape:
    tape.watch(x)
    l = square3(x)
grad3 = tape.gradient(l, x)
print(l)
print(grad3)
# >> tf.Tensor(100.0, shape=(), dtype=float32)
# >> tf.Tensor(0.0, shape=(), dtype=float32)

# Note: Gradients for square2 and square3 are 0, even though they are mathematically and (Python-) syntactically equivalent to square1's expected behavior
```

### Relevant log output

```shell

```

**Comment by Venkat6871** [comment]
I was able to reproduce the same issue on Colab using TensorFlow 2.18 and the nightly version. Please find the [gist](https://colab.sandbox.google.com/gist/Venkat6871/63e2a5aa4938e5446648f757cd15e43d/89110_tf_2-18-0-nightly-v.ipynb) attached for your reference.

Thank you!

**Comment by Samir-atra** [comment]
Hello,

I think the main topics to be addressed and checked here are:

- _Function scope:_ to check for local variables, global variables and how each one is being updated.
- _The floating precision_, and what changes when it gets passed from the function, or application of the cumulative and summative properties.
- _Mutating Data Passed by Reference_, and considering the different types of data structures when included in functions or calculated directly.

I hope this helps in reviewing the issue and fixing the error happening, or maybe clarifying what is happening in the process.

Kind regards
Samer

**Comment by Venkat6871** [comment]
Hi @adamvvu ,
Thank you so much for filing this issue. We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!

**Comment by Venkat6871** [comment]
Hi,
Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The observed difference is due to which tensor is passed to `tape.gradient()`. In the first example, x is reassigned to the clipped tensor before calling `tape.gradient(l, x)`, so the gradient is computed with respect to the clipped value (10.0) and returns `20.0.` In `square2` and `square3`, x still refers to the original input value (15.0), so the gradient through` tf.clip_by_value` is `0.0` outside the clipping range.

To compute the gradient with respect to the original input consistently, keep the original and clipped tensors as separate variables, for example:
`x_orig = tf.constant(15., dtype=tf.float32)`
`x_clipped = truncate_domain(x_orig, -10., 10.)`
`l = square1(x_clipped)`
`grad = tape.gradient(l, x_orig)`
This is expected behavior rather than an issue with defining the clipping operation inside a function.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.
Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by adamvvu** [comment]
Hi @Venkat6871 

Thanks for following up on this old issue. If it is expected behavior and not a bug, then this would imply the issue is simply that `tf.clip_by_value` **is not a differentiable operation**. This would be very important to keep in mind for those building models in TF. 

For example a very common actual use case may be:
```
x = # some input data/tensor
W = # weights of a NN

def safe_sigmoid(logits):
    clipped = tf.clip_by_value(logits, -10, 10)
    return sigmoid(clipped)

with tf.GradientTape() as tape:
    tape.watch(W)
    logits_hat = Wx
    p_hat = safe_sigmoid(logits_hat) # gradients break here
    loss = bce(Y, p_hat)
gradients = tape.gradient(loss, W)
```
I'm not on a computer so can't test it, but if it's expected behavior: implies `gradients` would be silently zero and halt all learning. Feel free to close issue if it is expected, and it may just a misconception that gradients are automatically clipped accordingly when using `tf.clip_by_value`.

**Comment by vruotolo** [comment]
Hi @adamvvu, regarding your `safe_sigmoid` snippet, this is actually working as intended, and the gradients "break" because of the mathematical properties of `tf.clip_by_value`.

The mathematical derivative of a clipping operation is exactly `0` whenever an input falls outside your bounds `([-10, 10])`. When `logits_hat` exceeds those bounds, `d(clipped)/d(logits_hat) = 0`. By the chain rule, this zeroes out the entire gradient for those weights and halts learning.

To prevent `NaN` values in BCE without manually clipping and killing your gradients, here are the two ways to solve this:

#### Solution A: The Idiomatic TensorFlow Way (Recommended)
The standard and most numerically stable approach is to use TensorFlow's built-in combined operation. It handles the numerical stability internally while providing perfect, non-zero gradients:

```python
with tf.GradientTape() as tape:
    tape.watch(W)
    logits_hat = W @ x
    
    # Automatically stable, computes sigmoid and BCE together, no clipping required
    loss = tf.nn.sigmoid_cross_entropy_with_logits(labels=Y, logits=logits_hat)
    
gradients = tape.gradient(loss, W)
```

#### Solution B: Custom Gradient (Straight-Through Estimator)
If you absolutely must use your custom `safe_sigmoid` function and separate bce loss for architectural reasons, you'll need to use `@tf.custom_gradient` to pass the gradients straight through the clipping operation during the backward pass:

```python
@tf.custom_gradient
def safe_sigmoid(logits):
    # Forward pass: perfectly clipped
    clipped = tf.clip_by_value(logits, -10., 10.)
    result = tf.math.sigmoid(clipped)
    
    # Backward pass: compute gradient as if the clip wasn't there
    def grad(dy):
        # gradient of sigmoid evaluated at the clipped value
        return dy * result * (1.0 - result)
        
    return result, grad
# ...
with tf.GradientTape() as tape:
    tape.watch(W)
    logits_hat = W @ x
    p_hat = safe_sigmoid(logits_hat) # Gradients will now flow perfectly here
    loss = bce(Y, p_hat)
gradients = tape.gradient(loss, W)
```

**Comment by dmiltr3** [comment]
Hi @adamvvu,

Closing this issue as this is working as intended based on the discussion above:

1. **Scoping / Variable Rebinding**: The discrepancy observed between the standalone and function definitions in the original example was due to Python variable rebinding. In `grad1`, `x = truncate_domain(x, ...)` rebound the variable `x` to the output tensor of the clip. Consequently, `tape.gradient(l, x)` computed the gradient with respect to the already-clipped tensor (value `10.0`), evaluating $\frac{d}{dx_{\text{clipped}}}(x_{\text{clipped}}^2) = 2 \times 10.0 = 20.0$ and bypassing the clipping op in the backward pass. In `square2` and `square3`, `x` remained bound to the original tensor (`15.0`), differentiating through `clip_by_value`.
2. **Autodiff & Derivative of Clipping**: For inputs outside the clipping range ($x < \text{lb}$ or $x > \text{ub}$), $f(x) = \text{clip}(x, \text{lb}, \text{ub})$ is constant, so its mathematical derivative is identically `0.0`. By the chain rule, any downstream gradient multiplied by this zero derivative becomes `0.0`.
3. **Recommended Solutions**: As highlighted in @vruotolo's comment above, to maintain numerical stability without killing gradients, we recommend using fused ops like `tf.nn.sigmoid_cross_entropy_with_logits` (Solution A) or implementing a straight-through estimator via `@tf.custom_gradient` (Solution B).

Thank you for opening the issue and for the discussion!
